Add options to `cache.gc` to enable deleting nonessential/recomputable result caching data

benjamn opened the pull request:
Ever since #3394, `InMemoryCache` has taken an approach of trading memory for improved efficiency of repeated cache reads.

More recently (as of #7439), this result caching system also guarantees cache results will be canonical, a technique that uses memory to store previous canonical objects in an efficient lookup structure (the `ObjectCanon`), to speed up future deep equality comparisons.

Although we continue to believe this memory/speed tradeoff is appropriate in most cases, it is conceivable that you might at some point need to reclaim memory more urgently than you need cache performance.

To release all cached result objects (to be garbage collected by JavaScript if otherwise unused) without resetting any actual cache data, this PR allows calling the `cache.gc` (#5310) method with a new `resetResultCache: true` option:
```ts
cache.gc({
  resetResultCache: true,
});
```
By default, resetting the result cache will not reset the `ObjectCanon`, which provides `===` continuity of cache results even after the result caching system has been completely replaced (as it is when you pass `resetResultCache: true` to `cache.gc`). That preservation is pretty nifty, but if you also want to dump the `ObjectCanon` for maximum memory savings, then you should also pass `resetResultIdentities: true` to `cache.gc`:
```ts
cache.gc({
  resetResultCache: true,
  resetResultIdentities: true,
});
```
These `cache.gc` options are available only on `InMemoryCache` (not on `ApolloCache`), since result caching is an `InMemoryCache`-specific concept.

The `resetResultCache` and `resetResultIdentities` options were the first two options that seemed useful to me, but we can easily add additional options in the future.

glasser commented on src/cache/inmemory/inMemoryCache.ts:
```diff
@@ -262,10 +276,21 @@ export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
-    return this.optimisticData.gc();
+    const ids = this.optimisticData.gc();
+    if (options && options.resetResultCache && !this.txCount) {
+      this.resetResultCache(options.preserveCanon);
```
Am I missing something, or does this fail to actually default to true?  Won't it call `this.resetResultCache(undefined)` which won't trigger the default value?

benjamn replied:
@glasser That's how it works! Default parameter values are triggered by undefined-ness, not (just) by omitting the arguments from the function call. This is useful especially when the parameter in question is not at the end of the parameter list.

glasser replied:
Wowzers, I had no idea! That does make sense though.

brainkim replied:
Hmmm I can imagine tripping up on this in the future.

brainkim replied:
I am aware of the `undefined` default parameter behavior but I can imagine tripping up on this in the future.

brainkim commented on package.json:
```diff
@@ -55,7 +55,7 @@
       "name": "apollo-client",
       "path": "./dist/apollo-client.cjs.min.js",
-      "maxSize": "24.35 kB"
+      "maxSize": "24.5 kB"
```
Should be lower if you rebase the branch.

brainkim commented on src/cache/inmemory/inMemoryCache.ts:
```diff
@@ -106,12 +106,26 @@ export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
+        canon: (
+          preserveCanon &&
+          previousReader &&
+          previousReader["canon"]
```
I would prefer making the “canon” public over this workaround to ignore private members.

benjamn replied:
`StoreReader` is a private API anyway, so this is fine with me.

benjamn replied:
https://github.com/apollographql/apollo-client/pull/8421/commits/13601395d614f3e8201f2ef3ac3ca21f39b99cde

brainkim commented on src/cache/inmemory/inMemoryCache.ts:
```diff
@@ -262,10 +280,21 @@ export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
+    resetResultCache?: boolean;
+    // If resetResultCache is true, this.storeReader.canon will be preserved by
+    // default, but can also be discarded by passing false for preserveCanon.
+    // Defaults to true, but has no effect if resetResultCache is false.
```
I am already having difficulty reasoning about the interaction between these two options based on the descriptions alone... I think part of the cognitive overload is that I expect both options to be moving in the same direction (both resetting or both preserving), and the other part is the use of a cascading default. Without reading documentation, I would expect `preserveCanon` to work independently. Like you can imagine wanting to make future entities which have been garbage collected to be non-equal referentially, while wanting to free as much internal memory as possible (`resetResultCache: true, preserveCanon: false` )?

benjamn replied:
The `{ resetResultCache: true, preserveCanon: false }` case is already supported. I think you're objecting to the case where only `preserveCanon: false` is provided. We could make `preserveCanon: false` dump the `ObjectCanon` by itself, without resetting the result cache?

benjamn replied:
https://github.com/apollographql/apollo-client/pull/8421/commits/21182663192608b21f455392a52d33eca64c7537

brainkim commented on src/cache/inmemory/inMemoryCache.ts:
```diff
@@ -262,10 +280,21 @@ export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
+    // If resetResultCache is true, this.storeReader.canon will be preserved by
+    // default, but can also be discarded by passing false for preserveCanon.
+    // Defaults to true, but has no effect if resetResultCache is false.
+    preserveCanon?: boolean;
```
I’m worried about exposing “canon” as a name to our users. Maybe we should run this by @StephenBarlow to see if this makes sense from the perspective of documentation. Are we exposing the “canon” name in public APIs elsewhere yet?

Another nit, it’s better to make boolean defaults false, insofar as `undefined` is falsy, and this allows you to avoid specificying a `true` ddefault in the `resetResultCache()` method.

Potential alternatives:

1. `resetX` where `X` is whatever public name we choose for referential equality stuff, as in `resetCanon` or `resetIdentical`
2. Maybe we could have an enumeration of levels of clearing? If preserveCanon is just another step beyond resetResultCache, creating levels of gc (`soft`, `hard`) might be appropriate.

benjamn replied:
How about `resetResultCache` and `resetResultIdentities`, both defaulting to `false`?

brainkim replied:
To be clear I’m also okay with using the `canon` name, just want it to be done intentionally.

benjamn replied:
https://github.com/apollographql/apollo-client/pull/8421/commits/21182663192608b21f455392a52d33eca64c7537

brainkim commented on src/cache/inmemory/readFromStore.ts:
```diff
@@ -127,12 +129,20 @@ export class StoreReader {
       addTypename: config.addTypename !== false,
     };
 
+    this.canon = config.canon || new ObjectCanon;
```
I think ObjectCanon parens shouldn’t be omitted `new ObjectCanon()`, but that’s just me.

brainkim commented on src/cache/inmemory/inMemoryCache.ts:
```diff
@@ -106,12 +106,26 @@ export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
+
+    // The StoreWriter is mostly stateless and so doesn't really need to be
+    // reset, but it does need to have its writer.storeReader reference updated,
+    // so it's simpler to update this.storeWriter as well.
```
Since this is a new addition to `gc()` logic, I wish there was a way to confirm that replacing both the `storeWriter` and `storeReader` made those two objects unreachable. That there isn’t some kind of secret chain retaining the old readers and writers.

Doesn’t have to be a unit test, just maybe opening Chrome DevTools and taking a heap snapshot before and after and searching for the old `StoreReader`/`StoreWriter` as a sanity check.

benjamn replied:
Good idea—we can use [`npm run test:memory`](https://github.com/apollographql/apollo-client/tree/release-3.4/scripts/memory) for this.

benjamn replied:
https://github.com/apollographql/apollo-client/pull/8421/commits/be7199b1828f0221f6e31cdd4ef88a47b662e515

brainkim reviewed (commented):
Wrote some thoughts. Again I’m not 100% familiar with how advanced caching stuff is supposed to work so my input is limited. My primary concerns are about the API design. Two new boolean options, which seem to be interrelated, and for an advanced use-case which might hard to search for/understand as a non-Apollo Client maintainer. Maybe we can loop in some advanced users like @sofianhn and the documentation team @StephenBarlow for their thoughts?

Also, if there are any relevant issues from users asking for this that I missed I’d love to see them as this could help with the design process.

benjamn commented:
@brainkim Thanks for the review! I think I've addressed your concerns, including writing a `FinalizationRegistry`-based test that was initially failing before I fixed the bug it caught, so thanks again for that reminder/suggestion.
